api/v1alpha4/clusterclass_webhook.go

```diff
@@ -91,77 +101,45 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
-				field.Invalid(
-					field.NewPath("spec", "workers", "machineDeployments", "template", "bootstrap", "ref", "namespace"),
-					class.Template.Bootstrap.Ref.Namespace,
-					"must match metadata.namespace",
-				),
-			)
-		}
-		if class.Template.Infrastructure.Ref != nil && class.Template.Infrastructure.Ref.Namespace != in.Namespace {
-			allErrs = append(allErrs,
-				field.Invalid(
-					field.NewPath("spec", "workers", "machineDeployments", "template", "infrastructure", "ref", "namespace"),
-					class.Template.Infrastructure.Ref.Namespace,
-					"must match metadata.namespace",
-				),
-			)
-		}
+	// Ensure all references are valid
+	allErrs = append(allErrs, in.validateAllRefs()...)
+
+	// Ensure all machineDeployment classes are unique
+	allErrs = append(allErrs, in.Spec.Workers.validateUniqueClasses(field.NewPath("spec", "workers"))...)
+
+	// Ensure spec changes are compatible
+	allErrs = append(allErrs, in.validateCompatibleSpecChanges(old)...)
+
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(GroupVersion.WithKind("ClusterClass").GroupKind(), in.Name, allErrs)
 	}
+	return nil
+}
 
-	// Ensure MachineDeployment class are unique.
-	classNames := sets.String{}
-	for _, class := range in.Spec.Workers.MachineDeployments {
-		if classNames.Has(class.Class) {
-			allErrs = append(allErrs,
-				field.Invalid(
-					field.NewPath("spec", "workers", "machineDeployments"),
-					class,
-					fmt.Sprintf("MachineDeployment class should be unique. MachineDeployment with class %q is defined more than once.", class.Class),
-				),
-			)
-		}
-		classNames.Insert(class.Class)
+func (in ClusterClass) validateAllRefs() field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, in.Spec.Infrastructure.validate(in.Namespace, field.NewPath("spec", "infrastructure"))...)
+	allErrs = append(allErrs, in.Spec.ControlPlane.LocalObjectTemplate.validate(in.Namespace, field.NewPath("spec", "controlplane"))...)
+	if in.Spec.ControlPlane.MachineInfrastructure != nil {
+		allErrs = append(allErrs, in.Spec.ControlPlane.MachineInfrastructure.validate(in.Namespace, field.NewPath("spec", "controlPlane", "machineInfrastructure"))...)
 	}
 
-	// in case of create, we are done.
-	if old == nil {
-		if len(allErrs) > 0 {
-			return apierrors.NewInvalid(GroupVersion.WithKind("ClusterClass").GroupKind(), in.Name, allErrs)
-		}
-		return nil
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		allErrs = append(allErrs, class.Template.Bootstrap.validate(in.Namespace, field.NewPath("spec", "workers", fmt.Sprintf("machineDeployments[%v]", i), "template", "bootstrap"))...)
+		allErrs = append(allErrs, class.Template.Infrastructure.validate(in.Namespace, field.NewPath("spec", "workers", fmt.Sprintf("machineDeployments[%v]", i), "template", "infrastructure"))...)
 	}
 
-	// Otherwise, In case of updates:
+	return allErrs
+}
+
+func (in ClusterClass) validateMachineDeploymentsChanges(old *ClusterClass) field.ErrorList {
```

Comment:
Can you move this func below under the func where it's called? Just easier to read (imho) top-down.